myhpi/tenca_django/mailinglist.py

```diff
@@ -0,0 +1,360 @@
+
+	def _raw_get_member(self, email):
+		try:
+			member = self.list.get_member(email)
+		except ValueError:
+			self._raise_nomember(email)
+		else:
+			return member
+
+	def add_member_silently(self, email):
+		"""Subscribe an email address to a mailing list, with no verification or notification send"""
+		self.list.subscribe(email, pre_verified=True, pre_confirmed=True, send_welcome_message=False)
+
+	def add_member(self, email, send_welcome_message=False, _on_retry=False):
+		"""Subscribes a user and sends them a confirmation mail
+
+		Returns the authentication token for that subscription.
+		"""
+		try:
+			return self.list.subscribe(email, send_welcome_message=send_welcome_message)["token"]
+		except urllib.error.HTTPError as e:
+			 # Already pending subscriptions are retried
+			if e.status == 409 and not _on_retry and settings.RETRY_CANCELS_PENDING_SUBSCRIPTION:
+				try:
+					token = next(t for t, e in self.pending_subscriptions().items() if e == email)
+				except StopIteration: # pragma: no cover
+					# This only happens in crazy race-conditions, if the request
+					# was accepted/canceled between with subscribe and lookup.
+					# If accepted, we expect again an error to notify.
+					# If canceled, the expected action is to correctly try again
+					pass
+				else:
+					self.cancel_pending_subscription(token)
+				return self.add_member(email, send_welcome_message, _on_retry=True)
+			else:
+				raise e
+
+	def toggle_membership(self, email):
+		"""Adds member if not present and removes if already member.
+		
+		This is useful from interfaces where you don't want to hint attackers
+		the current membership status of others.
+
+		Returns a tuple of (status, token), where status is True if the user was newly
+		added and False if it was removed.
+		"""
+		if self.is_member(email):
+			return False, self.remove_member(email)
+		else:
+			return True, self.add_member(email)
+
+
+	def configure_list(self):
+		self.list.settings.update(self.SHARED_LIST_DEFAULT_SETTINGS)
+		self.list.settings.update(settings.LIST_DEFAULT_SETTINGS)
+		self.list.settings['subject_prefix'] = '[{}] '.format(self.list.settings['list_name'].lower())
+		self.list.settings.save()
+
+	def configure_templates(self):
+		for mailman_template_name, tenca_template_name in self.TEMPLATE_MAPPINGS.items():
+			self.configure_single_template(mailman_template_name, tenca_template_name)
+
+	def configure_single_template(self, mailman_template_name, tenca_template_name):
+		template_args = dict(
+			fqdn_listname=self.fqdn_listname,
+			action_link=self.build_action_link('$token'),
+			action_abuse_link=self.build_action_abuse_link('$token'),
+			invite_link=self.build_invite_link(),
+			web_ui=urllib.parse.urljoin(settings.GET_SITE_URL(), reverse("tenca_dashboard"))
+		)
+		self.list.set_template(mailman_template_name, urllib.parse.urljoin(settings.SITE_URL, reverse("tenca_django:mailman_template", kwargs={'template_name': tenca_template_name})))
```

Comment:
Currently, the `template_args` are query parameters. However, mailman does not care about the URL format, so using a path-structure (e. g. `…/<fqdn-listname>/<action-link>/…`) or anything else would be okay, too. Old template links will break anyways and we have to change them all (as they point to `localhost:8080` and we're moving the templateserver to `myhpi.de`), so changing the URL format would not cause any additional effort.

Instead of inserting `fqdn_listname` into the domain on our own, we could also use the mailman placeholder `$listname` (<https://docs.mailman3.org/projects/mailman/en/latest/src/mailman/rest/docs/templates.html#url-placeholders>).